File: retort-annotations/src/main/lombok/co/kaioru/retort/annotation/builder/AnnotationFactoryBuilder.java
package co.kaioru.retort.annotation.builder;

import co.kaioru.retort.ICommand;
import co.kaioru.retort.ICommandContext;
import co.kaioru.retort.annotation.AnnotationFactory;
import co.kaioru.retort.annotation.IAnnotationAdapter;
import lombok.Getter;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

@Getter
public class AnnotationFactoryBuilder<I extends ICommandContext, O> extends AnnotationFactory<I, O> implements IAnnotationFactoryBuilder<I, O> {

    private final List<Object> objects;

    public AnnotationFactoryBuilder() {
        this.objects = new ArrayList<>();
    }

    @Override
    public IAnnotationFactoryBuilder<I, O> withAdapter(IAnnotationAdapter<I, O> adapter) {
        this.getPipeline().add(adapter);
        return this;
    }

    @Override
    public IAnnotationFactoryBuilder<I, O> withObject(Object object) {
        this.getObjects().add(object);
        return this;
    }

    @Override
    public Collection<ICommand<I, O>> build() {
        return getObjects().stream()
                .map(this::generate)
                .flatMap(Collection::stream)
                .collect(Collectors.toList());
    }

}

File: retort-annotations/src/main/kotlin/co/kaioru/retort/annotation/builder/impl/AnnotationFactoryBuilder.kt
package co.kaioru.retort.annotation.builder.impl

import co.kaioru.retort.ICommand
import co.kaioru.retort.annotation.builder.IAnnotationFactoryBuilder
import co.kaioru.retort.annotation.impl.AnnotationFactory

open class AnnotationFactoryBuilder<I, O> : AnnotationFactory<I, O>(), IAnnotationFactoryBuilder<I, O> {
    val objects: MutableCollection<Any> = HashSet()

    override fun withObject(any: Any): IAnnotationFactoryBuilder<I, O> {
        objects.add(any)
        return this
    }

    override fun withObjects(any: Collection<Any>): IAnnotationFactoryBuilder<I, O> {
        objects.addAll(any)
        return this
    }

    override fun build(): Collection<ICommand<I, O>> {
        return objects
                .flatMap { generate(it) }
                .toSet()
    }
}